### PR TITLE
[Merged by Bors] - Introduce VerifiedAtx as type

### LINKED ISSUE #3434: atxs: refactor atx type to avoid struct update
atx struct is updated in two cases:

- when received to compute id 
- and will be updated after it was verified, to set correct base_tick_height and tick_count https://github.com/spacemeshos/go-spacemesh/pull/3399

any mistakes can be catched only during runtime and only if we have sufficient test coverage. moreover this is prone to concurrency issues, if fields that are updated will be read concurrently in the other goroutine.

instead we should treat data as data and use different struct when we need more data. for example after atx is verified and verified data is required downstream we should use type:

```go
type VerifiedAtx struct {
     Atx
     BaseTickHeight uint64
     TickCount uint64
}
```

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation
Fixes https://github.com/spacemeshos/go-spacemesh/issues/3434

## Changes
- `ActivationTxHeader` and `ActivationTx` are now independent types instead of being nested. A header can still be generated from an ATX by calling the `Header()` method.
- `VerifiedActivationTxHeader` is a new type that is returned by calling `Verify` on an ATX.
  - Naming could be improved - open for suggestions
  - The method now ensures that the ID and Node ID fields are set, if not it calculates those
  - Additionally it requires the caller to specify `baseTickHeight` and `tickCount`
  - This type is now used everywhere where `baseTickHeight` and `tickCount` are expected to be set, which ensures compile time checks where previously errors would only happen during runtime

## Test Plan
All existing tests still pass

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: activation/handler.go [hunk 1]
```diff
@@ -139,7 +139,7 @@ func (h *Handler) ProcessAtx(ctx context.Context, atx *types.ActivationTx) error
 // SyntacticallyValidateAtx ensures the following conditions apply, otherwise it returns an error.
 //
 //   - If the sequence number is non-zero: PrevATX points to a syntactically valid ATX whose sequence number is one less
-//     than the current ATX's sequence number.
+//     than the current ATXs sequence number.
```
Comment (countvonzero) [inline activation/handler.go:142]: i don't think SequenceNumber used anymore. since you are here. can you update the comment so it make sense to you?

File: sql/atxs/atxs.go [hunk 2]
```diff
@@ -9,8 +9,8 @@ import (
 	"github.com/spacemeshos/go-spacemesh/sql"
 )
 
-// Get gets an ATX by a given ATX ID.
-func Get(db sql.Executor, id types.ATXID) (atx *types.ActivationTx, err error) {
+// ByID gets an ATX by a given ATX ID.
+func ByID(db sql.Executor, id types.ATXID) (atx *types.ActivationTx, err error) {
```
Comment (countvonzero) [inline sql/atxs/atxs.go]: `Get` is the consistent term in all the sql sub pkgs. i think we should either change all or leave it as Get

same for other methods. GetByXXXX

if you want change, please send a separate PR that updates all sql sub pkgs

**Review by countvonzero — COMMENTED** [review]
this PR is bigger than it should be.

please separate the renaming to a separate PR. and when you do, please make sure its consistent across pkgs inside sql.

**Comment by fasmat** [comment]
> this PR is bigger than it should be.
> 
> please separate the renaming to a separate PR. and when you do, please make sure its consistent across pkgs inside sql.

I'll undo most of the renaming and leave it for a different PR.

File: activation/handler.go (hunk 1 above)
Comment (fasmat) [reply activation/handler.go:142]: The method still behaves as described in the comment and the builder produces ATXs that pass this check, besides that I didn't see sequence number being used anywhere.

I will create a separate small PR that removes the Sequence number from the NIPostChallenge struct.

Comment (fasmat) [reply activation/handler.go:142]: PR can be found here: https://github.com/spacemeshos/go-spacemesh/pull/3520

Comment (countvonzero) [reply activation/handler.go:142]: oh. wait. i thought there was a field called SequenceNumber (as the comment indicated). 
but i don't know about the field Sequence tho. it may be still useful. i don't know. sorry for the confusion.

File: activation/handler.go [hunk 3]
```diff
@@ -112,7 +112,7 @@ func (h *Handler) ProcessAtx(ctx context.Context, atx *types.ActivationTx) error
 	h.processAtxMutex.Lock()
 	defer h.processAtxMutex.Unlock()
 
-	existingATX, _ := h.cdb.GetAtxHeader(atx.ID())
+	existingATX, _ := h.cdb.GetAtxByID(atx.ID())
```
Comment (evt) [inline activation/handler.go]: do we skip error here on purpose? 🤔

Comment (fasmat) [reply activation/handler.go]: We are only checking if the ATX with the given ID already exists in the cache / db. If it does, `existingATX` will be non-nil and `err` will be nil. If the ATX does not exist in the cache and cannot be fetched it will be `nil` and `err` will be non-nil.

The returned error will be some SQL error that would need to be parsed to be handled correctly. The only error that could give extra info besides "atx hasn't been stored yet" are DB connection issues. These arguably aren't needed to be handled here, since they will be handled a few lines further down if storing the ATX fails anyway.

File: common/types/activation.go [hunk 4]
```diff
@@ -317,6 +296,119 @@ func (atx *ActivationTx) GetShortPoetProofRef() []byte {
...
+}
+
+// SetNodeID sets the Node ID in the ATX's cache.
+func (atx *ActivationTx) SetNodeID(nodeID *NodeID) {
+	atx.nodeID = nodeID
+}
+
+// Verify an ATX for a given base TickHeight and TickCount.
+func (atx *ActivationTx) Verify(baseTickHeight, tickCount uint64) *VerifiedActivationTx {
+	// TODO(mafa): should we also invoke calcAndSetID() as well as calcAndSetNodeID() here? maybe after checking that all required fields are set?
```
Comment (fasmat) [inline common/types/activation.go]: Opinions? I could get rid of these two methods because in nearly all places the call chain looks like:

```golang
atx := newAtx()
atx.CalcAndSetID()
atx.CalcAndSetNodeID()
vAtx := atx.Verify(...)
```

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors try

File: activation/handler.go [hunk 5]
```diff
@@ -139,112 +139,112 @@ func (h *Handler) ProcessAtx(ctx context.Context, atx *types.ActivationTx) error
...
-		return fmt.Errorf("invalid nipost: %v", err)
+		return nil, fmt.Errorf("invalid nipost: %v", err)
 	}
-	atx.Verify(baseTickHeight, leaves/h.tickSize)
-	return nil
+
+	vAtx := atx.Verify(baseTickHeight, leaves/h.tickSize)
+
+	// TODO(mafa): instead of reporting the event immediately, should we only report it after validation passed?
+	events.ReportNewActivation(vAtx)
```
Comment (countvonzero) [inline activation/handler.go]: i don't know why it was reported immediately before, but i think the right thing to do is to report it after ProcessAtx() succeed, before "got new atx" logged.

File: common/types/activation.go [hunk 6]
```diff
@@ -180,18 +162,13 @@ func (atxh *ActivationTxHeader) TickHeight() uint64 {
...
-	atxh.tickCount = tickCount
-}
-
 // NIPostChallenge is the set of fields that's serialized, hashed and submitted to the PoET service to be included in the
 // PoET membership proof. It includes ATX sequence number, the previous ATX's ID (for all but the first in the sequence),
 // the intended publication layer ID, the PoET's start and end ticks, the positioning ATX's ID and for
 // the first ATX in the sequence also the commitment Merkle root.
 type NIPostChallenge struct {
+	// Sequence number counts the number of ancestors of the ATX. It sequentially increases for each ATX in the chain.
+	// It can be used in proving to other nodes that a malicious party created more ATXs than they would are allowed to.
```
Comment (countvonzero) [inline common/types/activation.go]: rephrase to "Two ATXs with the same sequence number from the same miner can be used as the proof of malfeasance against that miner."

File: common/types/activation.go [hunk 7]
```diff
@@ -223,9 +200,17 @@ func (challenge *NIPostChallenge) String() string {
 // structure is serialized and signed. It includes the header fields, as well as the larger fields that are only used
 // for validation: the NIPost and the initial Post.
 type InnerActivationTx struct {
```
Comment (countvonzero) [inline common/types/activation.go:192]: what is the motivation to remove the embedded-ness of the header?

File: common/types/activation.go [hunk 8]
```diff
@@ -274,11 +258,6 @@ func (atx *ActivationTx) MarshalLogObject(encoder log.ObjectEncoder) error {
 	encoder.AddUint32("epoch", uint32(atx.PubLayerID.GetEpoch()))
 	encoder.AddUint64("num_units", uint64(atx.NumUnits))
 	encoder.AddUint64("sequence_number", atx.Sequence)
-	if atx.verified {
-		encoder.AddUint64("base_tick_height", atx.baseTickHeight)
-		encoder.AddUint64("tick_count", atx.tickCount)
-		encoder.AddUint64("weight", atx.GetWeight())
```
Comment (countvonzero) [inline common/types/activation.go:280]: we should still log these. use atx.Header()?

File: common/types/activation.go (hunk 7 above)
Comment (countvonzero) [inline common/types/activation.go:192]: do you have a test that guarantee that InnerActivationTx has everything public field of ActivationTxHeader?
i.e. how do we keep these two struct in sync over time?

File: datastore/store.go [hunk 9]
```diff
@@ -51,13 +51,12 @@ func (db *CachedDB) GetAtxHeader(id types.ATXID) (*types.ActivationTxHeader, err
 		return nil, fmt.Errorf("get ATXs from DB: %w", err)
 	}
 
-	db.atxHdrCache.Add(id, &atx.ActivationTxHeader)
-	return &atx.ActivationTxHeader, nil
+	return atx.Header(), nil
```
Comment (countvonzero) [inline datastore/store.go]: here you are not getting it from the cache anymore. you are creating a new struct?

File: go.sum [hunk 10]
```diff
@@ -818,8 +818,8 @@ golang.org/x/net v0.0.0-20210726213435-c6fcb2dbf985/go.mod h1:9nx3DQGgdP8bBQD5qx
 golang.org/x/net v0.0.0-20220127200216-cd36cc0744dd/go.mod h1:CfG3xpIq0wQ8r1q4Su4UZFWDARRcnwPjda9FqA0JpMk=
 golang.org/x/net v0.0.0-20220225172249-27dd8689420f/go.mod h1:CfG3xpIq0wQ8r1q4Su4UZFWDARRcnwPjda9FqA0JpMk=
 golang.org/x/net v0.0.0-20220624214902-1bab6f366d9e/go.mod h1:XRhObCWvk6IyKnWLug+ECip1KBveYUHfp+8e9klMJ9c=
-golang.org/x/net v0.0.0-20220822230855-b0a4917ee28c h1:JVAXQ10yGGVbSyoer5VILysz6YKjdNT2bsvlayjqhes=
-golang.org/x/net v0.0.0-20220822230855-b0a4917ee28c/go.mod h1:YDH+HFinaLZZlnHAfSS6ZXJJ9M9t4Dl22yv3iI2vPwk=
+golang.org/x/net v0.0.0-20220826154423-83b083e8dc8b h1:ZmngSVLe/wycRns9MKikG9OWIEjGcGAkacif7oYQaUY=
+golang.org/x/net v0.0.0-20220826154423-83b083e8dc8b/go.mod h1:YDH+HFinaLZZlnHAfSS6ZXJJ9M9t4Dl22yv3iI2vPwk=
```
Comment (countvonzero) [inline go.sum]: why is this needed? shouldn't dependabot do this?

File: sql/atxs/atxs.go [hunk 11]
```diff
@@ -20,14 +20,14 @@ func Get(db sql.Executor, id types.ATXID) (atx *types.ActivationTx, err error) {
...
 		v.SetID(&id)
-		v.Verify(baseTickHeight, tickCount)
 		nodeID := types.NodeID{}
 		stmt.ColumnBytes(3, nodeID[:])
 		v.SetNodeID(&nodeID)
-		atx, err = &v, nil
+
+		baseTickHeight := uint64(stmt.ColumnInt64(1))
+		tickCount := uint64(stmt.ColumnInt64(2))
+		atx, err = v.Verify(baseTickHeight, tickCount), nil
```
Comment (countvonzero) [inline sql/atxs/atxs.go]: why ignore the error here?

File: common/types/activation_scale.go [hunk 12]
```diff
@@ -227,6 +256,29 @@ func (t *ActivationTx) DecodeScale(dec *scale.Decoder) (total int, err error) {
 	return total, nil
 }
 
+func (t *VerifiedActivationTx) EncodeScale(enc *scale.Encoder) (total int, err error) {
```
Comment (dshulyak) [inline common/types/activation_scale.go]: this needs to be removed

File: common/types/activation.go (hunk 8 above)
Comment (fasmat) [reply common/types/activation.go:280]: In an `ActivationTx` these values are always `0` because they have not been set. Only after calling `Verify` these values are set in the `VerifiedActivationTx`

Basically `atx.verified` is now guaranteed to be `false`.

Also `MarshalLogObject()` on a `VerifiedActivationTx` will log those values.

File: datastore/store.go (hunk 9 above)
Comment (fasmat) [reply datastore/store.go]: `GetAtxHeader` calles `GetFullAtx` if the header wasn't found in the cache. `GetFullAtx` already adds the header to the cache, no need to do it again here.

File: common/types/activation.go (hunk 7 above)
Comment (fasmat) [reply common/types/activation.go:192]: The header always has `baseTickHeight` and `tickCount` set, while an unverified ActivationTx does not. Making these two true separate objects removes the need for run time checks that these values are set.

You cannot get an `ActivationTxHeader` from an `ActivationTx` only from a `VerifiedActivationTx`.

File: go.sum (hunk 10 above)
Comment (fasmat) [reply go.sum]: `golang.org/x/net` and `google.golang.org/genproto` are libraries where google doesn't create new releases for - they don't tag them with a version number. These libraries can only be referenced by their commit hash. Every new commit is a bugfix or added feature though.

Dependabot doesn't update to unreleased versions, so as long as no version is tagged for these repos, dependabot won't update them automatically.

File: sql/atxs/atxs.go (hunk 11 above)
Comment (fasmat) [reply sql/atxs/atxs.go]: Which error? It's a double assignment:

```golang
atx = v.Verify(baseTickHeight, tickCount)
err = nil
```

before it was

```golang
v := functionCall()
atx, err = &v, nil
```

I'll reformat the code to make it clearer

File: common/types/activation_scale.go (hunk 12 above)
Comment (fasmat) [reply common/types/activation_scale.go]: This is generated by `go generate`. The only way to remove `EncodeScale` and `DecodeScale` for `VerifiedActivationTx` is to move them to a separate file, which I now did.

File: common/types/activation.go [hunk 13]
```diff
@@ -317,6 +296,54 @@ func (atx *ActivationTx) GetShortPoetProofRef() []byte {
...
+	atx.id = id
+}
+
+// SetNodeID sets the Node ID in the ATX's cache.
+func (atx *ActivationTx) SetNodeID(nodeID *NodeID) {
+	atx.nodeID = nodeID
+}
+
+// Verify an ATX for a given base TickHeight and TickCount.
+func (atx *ActivationTx) Verify(baseTickHeight, tickCount uint64) *VerifiedActivationTx {
```
Comment (poszu) [inline common/types/activation.go]: This method's comment and name imply that this function performs **verification**. But it only constructs a "verified" type assuming the caller has already verified it somehow. It's misleading. Perhaps rename the method to something like `ToVerified`.

File: common/types/verified_activation.go [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+	return vatx.tickCount
+}
+
+// TickHeight returns a sum of base tick height and tick count.
+func (vatx *VerifiedActivationTx) TickHeight() uint64 {
+	return vatx.baseTickHeight + vatx.tickCount
+}
+
+// Header returns the header of the atx.
+func (vatx *VerifiedActivationTx) Header() *ActivationTxHeader {
```
Comment (poszu) [inline common/types/verified_activation.go]: What's the motivation to return a pointer here instead of just a value? The users of this API might think that this method returns a pointer to the insides of `vatx`, but it returns a separate object (a copy basically). And they must consider what happens if `nil` is returned (I know it cannot return a `nil` but to know that you need to read the implementation).

I'd suggest returning by value. What do you think?

File: common/types/verified_activation.go [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+type VerifiedActivationTx struct {
+	*ActivationTx
+
+	baseTickHeight uint64
+	tickCount      uint64
+}
+
+// GetWeight of the atx.
+func (vatx *VerifiedActivationTx) GetWeight() uint64 {
+	return uint64(vatx.NumUnits) * (vatx.tickCount)
```
Comment (poszu) [inline common/types/verified_activation.go]: This might be a stupid question... I don't know the practical bounds of these variables yet :)

Can it, in practice, overflow here? It's a uint32 * uint64. Perhaps consider annotating in a comment why it is safe to multiply.

Similarly when you add two `uint64` in `TickHeight()`.

File: common/types/activation.go (hunk 13 above)
Comment (fasmat) [reply common/types/activation.go]: While true, I avoided doing too many changes in one PR; the method already existed before and did the same thing (except for setting fields in `ActivationTx` instead of returning a new `VerifiedActivationTx`). I can rename it to `ToVerifiedActivationTx(baseTickHeight, tickCount uint64)`.

@dshulyak I also added a TODO in this method. I think ID and Node ID should be calculated here since I would expect a `VerifiedActivationTx` to have these two fields set. At the moment this isn't guaranteed and adding those would to some degree justify calling this method `Verify`.

File: common/types/activation.go (hunk 7 above)
Comment (fasmat) [reply common/types/activation.go:192]: I don't have a test, I'm also not sure how such a test would look like.

The only way to create an `ActivationTxHeader` (besides calling `new(ActivationTxHeader)`) is to call the method `VerifiedActivationTx.Header()`. So any field that a Header has needs to be set by this method from a field of a `VerifiedActivationTx` which includes `InnerActivationTx`.

File: common/types/verified_activation.go (hunk 14 above)
Comment (fasmat) [reply common/types/verified_activation.go]: You are correct. `Header()` can never return `nil` not even conceptually, since every ATX has a Header. The main reason is again avoiding too many changes at once.

`Header()` returns a pointer to the object because everywhere we receive an `ActivationTxHeader` we expect a pointer instead of a value, i.e. https://github.com/spacemeshos/go-spacemesh/blob/store-and-chacke-full-atxs/activation/handler.go#L125 If header returns a value instead of a reference I would need to either change `ContextuallyValidateAtx` or change this code to something like:

```golang
header := atx.Header()
h.ContextuallyValidateAtx(&header)
```

Comment (fasmat) [reply common/types/verified_activation.go]: Also returning it as a value wouldn't be enough. `NIPostChallenge` contains a byte slice `InitialPostIndices` that would require a copy as well to make it a truly separate object.

I'm not sure about the size of this array, but if I remember correctly a Header is expected to be in the range of ~ 100 byte in size for the main net.

File: common/types/verified_activation.go (hunk 15 above)
Comment (fasmat) [reply common/types/verified_activation.go]: It's a valid point. I'm not sure about the expected size of these variables as well. The biggest I have seen where < 1000, which after multiplication would still easily fit into a `uint64`, but maybe @dshulyak can give some insights here? :-)

File: sql/atxs/atxs.go [hunk 16]
```diff
@@ -15,7 +15,7 @@ func Get(db sql.Executor, id types.ATXID) (atx *types.VerifiedActivationTx, err
 		stmt.BindBytes(1, id.Bytes())
 	}
 	dec := func(stmt *sql.Statement) bool {
-		var v types.VerifiedActivationTx
```
Comment (countvonzero) [inline sql/atxs/atxs.go]: there should be a test that failed before your fix but passed with this.
i.e. 
- add a atx to db
- get the blob out
- parse it correctly

**Review by countvonzero — COMMENTED** [review]
please also address questions in the earlier review. thanks

File: datastore/store.go (hunk 9 above)
Comment (countvonzero) [reply datastore/store.go]: it adds to the cache in `GetFullAtx`, but you are not returning the cached value. you are creating a new struct. what am i missing?

Comment (countvonzero) [reply datastore/store.go]: also, what is the value of the header cache if one can arbitrarily call atx.Header() now?
should Header() be a private method in the cache instead?

File: go.sum (hunk 10 above)
Comment (countvonzero) [reply go.sum]: but this change is unrelated right?
i've had the unpleasant experience of binary search on a failed deps update along with feature change...

File: common/types/verified_activation.go (hunk 14 above)
Comment (countvonzero) [reply common/types/verified_activation.go]: my understanding was that headers are always cached, hence the ptr

File: datastore/store.go (hunk 9 above)
Comment (fasmat) [reply datastore/store.go]: I didn't change any behaviour. The code behaves the same as before:

- If you call either `GetFullAtx(id)` or `GetAtxHeader(id)` you will receive the Atx / header for the given ID
- If `GetAtxHeader(id)` is called and the value is cached, the cached value is returned
- If `GetFullAtx(id)` or `GetAtxHeader(id)` are called when the header for the given ID is not cached it will be cached

The only thing that changed is: if a header isn't cached for a given ID it is now only added to the cache once instead of twice.

File: common/types/activation.go (hunk 4 above)
Comment (countvonzero) [reply common/types/activation.go]: i am in favor if you can make it work.

File: datastore/store.go (hunk 9 above)
Comment (fasmat) [reply datastore/store.go]: > also, what is the value of the header cache if one can arbitrarily call atx.Header() now? should Header() be a private method in the cache instead?

It's the same as before: one could have arbitrarily called `atx.ActivationTxHeader` instead of fetching the header from the cache and `Header()` can't be a private method because it is required to be called in packages other than `common/types`

Comment (countvonzero) [reply datastore/store.go]: not what my reading.
old code:
```
func (db *CachedDB) GetAtxHeader(id types.ATXID) (*types.ActivationTxHeader, error) {
	if id == *types.EmptyATXID {
		return nil, errors.New("trying to fetch empty atx id")
	}

	if atxHeader, gotIt := db.atxHdrCache.Get(id); gotIt {
		return atxHeader, nil
	}
	atx, err := db.GetFullAtx(id)
	if err != nil {
		return nil, fmt.Errorf("get ATXs from DB: %w", err)
	}

	db.atxHdrCache.Add(id, &atx.ActivationTxHeader)
	return &atx.ActivationTxHeader, nil
}
```
it cached the same ptr twice (once in `GetFullAtx()`, and once in `GetAtxHeader()`)
but it is returning the cached ptr. you are now returning a new struct.

File: go.sum (hunk 10 above)
Comment (fasmat) [reply go.sum]: If this PR gets merged first: https://github.com/spacemeshos/go-spacemesh/pull/3530 a rebase will get rid of this change, but for some reason bors seems to not work for the last few hours...

File: datastore/store.go (hunk 9 above)
Comment (countvonzero) [reply datastore/store.go]: > It's the same as before: one could have arbitrarily called atx.ActivationTxHeader instead of fetching the header from the cache and Header() can't be a private method because it is required to be called in packages other than common/types

no. before all the header ptrs are cached after retrieving the full atx from db. here you are creating a new struct every time Header() is called.

File: common/types/verified_activation.go (hunk 14 above)
Comment (fasmat) [reply common/types/verified_activation.go]: The change I made is instead of calling `header := &atx.ActivationTxHeader` one now calls `header := atx.Header()` unless one explicitly fetches the ATX or its header from the store it will NOT be cached and wasn't cached before.

File: datastore/store.go (hunk 9 above)
Comment (fasmat) [reply datastore/store.go]: I get your point, but what is the behaviour we are expecting?

```golang
id := 1234
header1 := db.GetAtxHeader(id)

// make some change
header1.PrevATXID = 5678

header2 := db.GetAtxHeader(id)
same := header1 == header2
equal := reflect.DeepEqual(header1, header2)
```

Is `same` and `equal` expected to be `true` or `false`? Because before it could be either - `true` as long as the value was cached (last 600 entries fetched from store) and `false` if the cache became invalid between the two calls to `db.GetAtxHeader`.

Comment (countvonzero) [reply datastore/store.go]: i don't understand your example. i don't think that's how the codebase use the atx header cache....

the expected behavior is caching the header, not creating more memory than we need to.
now if you want to argue the need of the cache, that's a different matter.

Comment (fasmat) [reply datastore/store.go]: I removed the `Header()` method from `VerifiedActivationTx`. Now a header can only be fetched from the store/cache and is cached correctly after fetching it (or the full ATX with the same ID).

File: sql/atxs/atxs.go (hunk 16 above)
Comment (fasmat) [reply sql/atxs/atxs.go]: There is a test in `store_test.go` that asserts this: `TestBlobStore_GetATXBlob`

Comment (countvonzero) [reply sql/atxs/atxs.go]: but why didn't it fail when systest failed when the wrong type was parsed?

**Comment by fasmat** [comment]
bors try

File: activation/handler.go [hunk 17]
```diff
@@ -139,117 +139,116 @@ func (h *Handler) ProcessAtx(ctx context.Context, atx *types.ActivationTx) error
...
-			return fmt.Errorf("validation failed: prevATX not found: %v", err)
+			return nil, fmt.Errorf("validation failed: prevATX not found: %v", err)
 		}
 
-		if prevATX.NodeID() != atx.NodeID() {
-			return fmt.Errorf("previous atx belongs to different miner. atx.ID: %v, atx.NodeID: %v, prevAtx.NodeID: %v",
-				atx.ShortString(), atx.NodeID(), prevATX.NodeID())
+		if prevATX.NodeID != atx.NodeID() {
+			return nil, fmt.Errorf(
+				"previous atx belongs to different miner. atx.ID: %v, atx.NodeID: %v, prevAtx.NodeID: %v",
```
Comment (countvonzero) [inline activation/handler.go]: can you also print out prevAtx's ID short string? thanks

File: api/grpcserver/grpcserver_test.go [hunk 18]
```diff
@@ -114,10 +116,33 @@ var (
...
 
+	atx := types.NewActivationTx(challenge, addr1, nipost, numUnits, nil)
+	if err := activation.SignAtx(signer, atx); err != nil {
+		log.Println("failed to sign atx:", err)
+		os.Exit(1)
+	}
+	var err error
+	globalAtx, err = atx.Verify(0, 1)
+	if err != nil {
+		log.Println("failed to verify atx", err)
```
Comment (countvonzero) [inline api/grpcserver/grpcserver_test.go]: does the err needs formatting?

File: common/types/verified_activation.go [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+type VerifiedActivationTx struct {
+	*ActivationTx
+
+	baseTickHeight uint64
+	tickCount      uint64
+}
+
+// GetWeight of the atx.
+func (vatx *VerifiedActivationTx) GetWeight() uint64 {
```
Comment (countvonzero) [inline common/types/verified_activation.go:18]: the result should always be the same as ActivationTxHeader. maybe worth factoring this out to a func that takes unit and tickcount. similarly for TickHeight.

**Review by countvonzero — DISMISSED** [review]
nicely done. thanks

File: api/grpcserver/grpcserver_test.go (hunk 18 above)
Comment (fasmat) [reply api/grpcserver/grpcserver_test.go]: TIL: `fmt.Println` (and `log` is just an extension of that) will not only check if an object implements the `fmt.Stringer` interface, but also `error` and automatically call `err.Error()` to get the string for err. So if just the message needs to be printed, no extra formatting is needed.

proof: https://go.dev/play/p/Jd5JrIsVVQc

File: common/types/activation.go [hunk 20]
```diff
@@ -119,9 +126,24 @@ func (atxh *ActivationTxHeader) TargetEpoch() EpochID {
 	return atxh.PubLayerID.GetEpoch() + 1
 }
 
-// GetWeight of the atx.
+// GetWeight of the ATX. The total weight of the epoch is expected to fit in a uint64 and is
+// sum(atx.NumUnits * atx.TickCount for each ATX in a given epoch).
+// Space Units sizes are chosen such that NumUnits for all ATXs in an epoch is expected to be < 10^6.
```
Comment (noamnelke) [inline common/types/activation.go]: `10^6` for all ATXs is wrong. Using Chia as a reference, their historical maximum netspace was >36 EiB which is 151,000,000 space units of 256 GiB. So we should expect / be prepared for `10^9`.

File: common/types/activation.go [hunk 21]
```diff
@@ -119,9 +126,24 @@ func (atxh *ActivationTxHeader) TargetEpoch() EpochID {
...
+	return safeMul(uint64(atxh.NumUnits), (atxh.TickCount))
+}
+
+func safeMul(a, b uint64) uint64 {
+	if a > 0 && b > math.MaxUint64/a {
+		panic("uint64 overflow")
+	}
+	if b > 0 && a > math.MaxUint64/b {
+		panic("uint64 overflow")
+	}
```
Comment (noamnelke) [inline common/types/activation.go]: Why is the second check needed? I assume it's because of rounding down that happens in the division. You could do one check if you did (on line 140) `b >= math.MaxUint64/a + 1`, but all of this is actually really confusing and error prone...

I find [Rob Pike's version](https://groups.google.com/g/golang-nuts/c/h5oSN5t3Au4/m/KaNQREhZh0QJ) with a single division more intuitively obvious - I need to think less to see that it's safe.

Adapted here it would be:
```go
func safeMul(a, b uint64) uint64 {
	c := a * b
	if a > 1 && b > 1 && c/b != a {
		panic("uint64 overflow")
	}
	return c
}
```

I'd be curious to see a benchmark comparison of the 3 options (and also to make sure that my version passes tests - maybe I'm missing something).

**Review by noamnelke — CHANGES_REQUESTED** [review]
(second comment is optional - ignore if you disagree)

File: common/types/activation.go (hunk 21 above)
Comment (fasmat) [reply common/types/activation.go]: A very quick benchmark shows that this is not only faster than the double check version, but also hardly slower than multiplying without overflow check:

```
goos: linux
goarch: arm64
pkg: github.com/spacemeshos/go-spacemesh/common/types
BenchmarkSafeMul-6    	940181595	         1.256 ns/op	       0 B/op	       0 allocs/op
BenchmarkSafeMul2-6   	1000000000	         0.6292 ns/op	       0 B/op	       0 allocs/op
BenchmarkMul-6   	1000000000	         0.4933 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/spacemeshos/go-spacemesh/common/types	2.459s
```

I'll update the implementation to this version.

**Comment by fasmat** [comment]
bors try

**Review by poszu — DISMISSED** [review]
My comments have been addressed. LGTM 👍

**Review by noamnelke — DISMISSED** [review]

**Comment by fasmat** [comment]
bors merge

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
bors merge

**Comment by countvonzero** [comment]
@fasmat i think i prefer to up the overall run tests time to 70min, and leave the systest part as 50min.

File: common/types/activation.go [hunk 22]
```diff
@@ -133,7 +133,11 @@ func (atxh *ActivationTxHeader) TargetEpoch() EpochID {
 // to produce more over time. A uint64 should be large enough to hold the total weight of an epoch,
 // for at least the first few years.
 func (atxh *ActivationTxHeader) GetWeight() uint64 {
-	return safeMul(uint64(atxh.NumUnits), (atxh.TickCount))
+	return getWeight(uint64(atxh.NumUnits), (atxh.TickCount))
```
Comment (countvonzero) [inline common/types/activation.go]: nit: (atxh.TickCount) the parenthesis not needed.

File: common/types/verified_activation.go [hunk 23]
```diff
@@ -9,9 +9,14 @@ type VerifiedActivationTx struct {
...
-// GetWeight of the atx.
+// GetWeight of the ATX. The total weight of the epoch is expected to fit in a uint64 and is
+// sum(atx.NumUnits * atx.TickCount for each ATX in a given epoch).
+// Space Units sizes are chosen such that NumUnits for all ATXs in an epoch is expected to be < 10^6.
+// PoETs should produce ~10k ticks at genesis, but are expected due to technological advances
+// to produce more over time. A uint64 should be large enough to hold the total weight of an epoch,
+// for at least the first few years.
 func (vatx *VerifiedActivationTx) GetWeight() uint64 {
-	return uint64(vatx.NumUnits) * (vatx.tickCount)
+	return getWeight(uint64(vatx.NumUnits), (vatx.tickCount))
```
Comment (countvonzero) [inline common/types/verified_activation.go]: ditto

**Comment by fasmat** [comment]
bors try

**Comment by fasmat** [comment]
> @fasmat i think i prefer to up the overall run tests time to 70min, and leave the systest part as 50min.

It works but is extremely close - the systests finished in 49 minutes; I would feel more comfortable if we would increase the timeout to 60 min.

EDIT: I decreased the timeout of the run tests step to 50 minutes (from 60) and increased the timeout for the systest job from 50 to 60 minutes

**Review by countvonzero — APPROVED** [review]

**Comment by fasmat** [comment]
bors merge
